File: 0322-coin-change/0322-coin-change.java
class Solution {
    public int coinChange(int[] coins, int amount) {
        return change(coins, amount, new int[amount + 1]);
    }

    public int change(int[] coins, int value, int[] memo) {
        if (value < 0) return -1;
        if (value == 0) return 0;
        if (memo[value] != 0) return memo[value];

        int min = Integer.MAX_VALUE;
        for (int i = 0; i < coins.length; ++i) {
            int ans = change(coins, value - coins[i], memo);
            if (ans >= 0 && ans < min) {
                min = 1 + ans;
            }
        }

        return memo[value] = min == Integer.MAX_VALUE ? -1 : min;
    }
}

File: 0322-coin-change/0322-coin-change.py
class Solution:
    ''' 
    dp[s] = min(dp[s], 1 + dp[s - coin])
    '''
    def coinChange(self, coins: List[int], amount: int) -> int:
        dp = [float('inf') for _ in range(amount + 1)]
        dp[0] = 0
        for s in range(amount + 1):
            for coin in coins:
                if s >= coin:
                    dp[s] = min(dp[s], 1 + dp[s - coin])
        return -1 if dp[amount] == float('inf') else dp[amount]
        